### PR TITLE
doc: tiered proxy urls

### LINKED ISSUE #311: Document tiered proxies

### PR DESCRIPTION
**Pull request by janbuchar**
- closes #311

### CODE REVIEW

File: docs/guides/proxy_management.mdx [hunk 1]
```diff
@@ -81,6 +84,25 @@ When no `session_id` is provided, our proxy URLs are rotated round-robin.
     </TabItem>
 </Tabs>
 
+### Tiered proxies
+
+When you use HTTP proxies in real world crawling scenarios, you have to decide which type of proxy to use to reach a sweetspot between cost efficiency and reliably avoiding blocking. Some websites may allow crawling with no proxy, on some you may get away with using datacenter proxies, which are cheap but likely to get blocked, and sometimes you need to use expensive residential proxies.
```
Comment (barjin) [inline docs/guides/proxy_management.mdx:89]:
```suggestion
When you use HTTP proxies in real world crawling scenarios, you have to decide which type of proxy to use to reach the sweet spot between cost efficiency and reliably avoiding blocking. Some websites may allow crawling with no proxy, on some you may get away with using datacenter proxies, which are cheap but likely to get blocked, and sometimes you need to use expensive residential proxies.
```

File: docs/guides/proxy_management.mdx [hunk 2]
```diff
@@ -81,6 +84,25 @@ When no `session_id` is provided, our proxy URLs are rotated round-robin.
...
+
+To take the guesswork out of this process, Crawlee allows you to configure multiple tiers of proxy URLs. When crawling, it will automatically pick the left-most tier where it doesn't encounter blocking. If you organize your proxy server URLs in tiers so that the left-most tier contains the cheapest, least reliable ones and each higher tier contains more expensive, more reliable ones, you will get an optimal anti-blocking performance.
```
Comment (barjin) [inline docs/guides/proxy_management.mdx:91]: I'm fine with it either way, but I always thought of tiers as top-down layers - so the first one is the "lowest", while the last one is the "top tier" (this is afaik an actual English phrase).

File: docs/guides/proxy_management.mdx [hunk 3]
```diff
@@ -81,6 +84,25 @@ When no `session_id` is provided, our proxy URLs are rotated round-robin.
...
+
```
Comment (barjin) [inline docs/guides/proxy_management.mdx:92]: Also, maybe mention that the proxies in the active tier are rotated round robin, the 2D array of URLs can get confusing (my fault 😅)

File: src/crawlee/proxy_configuration.py
Comment (barjin) [inline src/crawlee/proxy_configuration.py:1]: You sure this should be here? just surprised to see actual code change in a docs PR, so making sure this is intended :)

**Review by barjin — APPROVED** [review]
More or less nits, lgtm otherwise 👍 🎉

File: src/crawlee/proxy_configuration.py
Comment (janbuchar) [reply src/crawlee/proxy_configuration.py:1]: It is foul play, but I saw an unused property while adding a docblock and just removed it. We're pre-1.0, yolo.

**Review by vdusek — APPROVED** [review]
LGTM

File: docs/guides/proxy_management.mdx [hunk 4]
```diff
@@ -81,6 +84,25 @@ When no `session_id` is provided, our proxy URLs are rotated round-robin.
     </TabItem>
 </Tabs>
 
+### Tiered proxies
+
+When you use HTTP proxies in real world crawling scenarios, you have to decide which type of proxy to use to reach the sweet spot between cost efficiency and reliably avoiding blocking. Some websites may allow crawling with no proxy, on some you may get away with using datacenter proxies, which are cheap but likely to get blocked, and sometimes you need to use expensive residential proxies.
```
Comment (B4nan) [inline docs/guides/proxy_management.mdx:89]: > which are cheap but likely to get blocked

maybe better say they are easily detected instead? idk

File: docs/guides/proxy_management.mdx (hunk 2 above)
Comment (B4nan) [reply docs/guides/proxy_management.mdx:91]: yeah agreed

File: src/crawlee/proxy_configuration.py [hunk 5]
```diff
@@ -113,9 +112,19 @@ def __init__(
...
     ) -> None:
-        self._configuration = configuration or Configuration.get_global_configuration()
+        """Initialize a proxy configuration object.
+
+        Exactly one of `proxy_urls`, `tiered_proxy_urls` or `new_url_function` must be specified.
+
+        Args:
+            proxy_urls: A list of URLs of proxies that will be rotated in a round-robin fashion
+            tiered_proxy_urls: A list of URL tiers (where a tier is a list of proxy URLs). Crawlers will automatically
+                try to use the left-most (lowest index) tier where blocking does not happen. The proxy URLs in
```
Comment (B4nan) [inline src/crawlee/proxy_configuration.py:123]: this also uses left/right instead of top/bottom

File: docs/guides/proxy_management.mdx (hunk 2 above)
Comment (janbuchar) [reply docs/guides/proxy_management.mdx:91]: okay, so higher index ~ higher tier? I'll probably explain the relation of high/low tier to the index in the array anyway

Comment (B4nan) [reply docs/guides/proxy_management.mdx:91]: yeah, higher is more expensive and better working, that way it feels relatively understandable :D

**Review by barjin — APPROVED** [review]
👍

**Pull request merged**